### PR TITLE
feat: add work item button component

### PR DESCRIPTION
**Pull request by Yelinz**
This component takes a work item id and state keyword to call the appropriate state mutation on the given work item

Next step would be to create a "TaskButton" which expects a task slug and filtering to fetch the work item which will be mutated

### CODE REVIEW

File: addon/components/work-item-button.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  @queryManager apollo;
+
+  @service intl;
+
+  constructor(owner, args) {
+    super(owner, args);
+
+    this.cancelWorkItem = cancelWorkItem;
+    this.completeWorkItem = completeWorkItem;
+    this.skipWorkItem = skipWorkItem;
```
Comment (anehx) [inline addon/components/work-item-button.js:20]: You can assign those as properties directly, no need for a constructor. Also, postfix them with `Mutation`:

```js
cancelWorkItemMutation = cancelWorkItem;
// ...
```

File: addon/components/work-item-button.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { inject as service } from "@ember/service";
+import Component from "@glimmer/component";
+import { queryManager } from "ember-apollo-client";
+import { restartableTask } from "ember-concurrency-decorators";
+
+import cancelWorkItem from "ember-caluma/gql/mutations/cancel-work-item.graphql";
+import completeWorkItem from "ember-caluma/gql/mutations/complete-work-item.graphql";
+import skipWorkItem from "ember-caluma/gql/mutations/skip-work-item.graphql";
+
+export default class ComponentsWorkItemButtonComponent extends Component {
```
Comment (anehx) [inline addon/components/work-item-button.js:10]:
```suggestion
export default class WorkItemButtonComponent extends Component {
```

File: addon/components/work-item-button.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  constructor(owner, args) {
+    super(owner, args);
+
+    this.cancelWorkItem = cancelWorkItem;
+    this.completeWorkItem = completeWorkItem;
+    this.skipWorkItem = skipWorkItem;
+  }
+
+  @restartableTask
```
Comment (anehx) [inline addon/components/work-item-button.js:23]: Should be `dropTask` since restarting such a mutation would not make sense. A work item can only be skipped, completed or canceled once

File: addon/templates/components/work-item-button.hbs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<UkButton
+  ...attributes
+  @type={{@type}}
+  @disabled={{@disabled}}
+  @active={{@active}}
+  @loading={{@loading}}
```
Comment (anehx) [inline addon/templates/components/work-item-button.hbs:6]:
```suggestion
  @loading={{or @loading this.mutate.isRunning}}
```

File: addon/templates/components/work-item-button.hbs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<UkButton
+  ...attributes
+  @type={{@type}}
+  @disabled={{@disabled}}
```
Comment (anehx) [inline addon/templates/components/work-item-button.hbs:4]:
```suggestion
  @disabled={{or @disabled this.mutate.isRunning}}
```

File: addon/templates/components/work-item-button.hbs [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  @active={{@active}}
+  @loading={{@loading}}
+  @color={{or @color "default"}}
+  @size={{@size}}
+  @title={{@title}}
+  @on-click={{perform this.mutate}}
+>{{t (concat "caluma.mutate-work-item." @mutation)}}</UkButton>
```
Comment (anehx) [inline addon/templates/components/work-item-button.hbs:11]:
```suggestion
>
  {{#if (has-block)}}
    {{yield}}
  {{else if @label}}
    {{@label}}
  {{else}}
    {{t (concat "caluma.mutate-work-item." @mutation)}}
  {{/if}}
</UkButton>
```

File: tests/integration/components/work-item-button-test.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  setupRenderingTest(hooks);
+  setupIntl(hooks);
+
+  test("it renders", async function (assert) {
+    await render(
+      hbs`<WorkItemButton @mutation="complete" @workItemId="test" />`
+    );
+
+    assert.dom(this.element).hasText("");
+  });
```
Comment (anehx) [inline tests/integration/components/work-item-button-test.js:17]: Needs proper tests

File: translations/fr.yaml [hunk 8]
```diff
@@ -73,3 +73,8 @@ caluma:
       delete: "Supprimer l'option"
       archive: "Archiver (masquer) l'option"
       restore: "Restaurer l'option"
+
+  mutate-work-item:
+    complete: "Compléter"
```
Comment (anehx) [inline translations/fr.yaml:78]:
```suggestion
    complete: "Terminer"
```

File: translations/fr.yaml [hunk 9]
```diff
@@ -73,3 +73,8 @@ caluma:
...
+    skip: "Passer"
```
Comment (anehx) [inline translations/fr.yaml:79]:
```suggestion
    skip: "Sauter"
```

**Review by anehx — CHANGES_REQUESTED** [review]

**Comment by anehx** [comment]
Some basic documentation of the component should be added as well

**Comment by Yelinz** [comment]
Added some documentation and applied all changes.
I'm having some problems with the tests which I'll try to figure out.

While writing the docs, i noticed that glimmer components aren't all too compatible with the yuidoc style as you cant document the arguments without defining them. But I might be wrong there.

File: addon/templates/components/work-item-button.hbs [hunk 10]
```diff
@@ -0,0 +1,19 @@
+<UkButton
+  ...attributes
+  @type={{@type}}
+  @disabled={{or @loading this.mutate.isRunning}}
```
Comment (anehx) [inline addon/templates/components/work-item-button.hbs:4]:
```suggestion
  @disabled={{or @disabled @loading this.mutate.isRunning}}
```

File: tests/integration/components/work-item-button-test.js (hunk 7 above)
Comment (anehx) [reply tests/integration/components/work-item-button-test.js:33]: Test the mutations please

File: addon/components/work-item-button.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  @queryManager apollo;
+
+  @service intl;
+
+  /**
+   * The mutation which is going to be used for the work item.
+   * Valid mutations are: complete, skip, cancel
+   * @argument {String} mutation
+   */
+  mutation;
```
[the remark this answers is not in the document]
Comment (Yelinz) [reply addon/components/work-item-button.js]: You're right, they just need to be in separate comments

File: addon/templates/components/work-item-button.hbs (hunk 10 above)
Comment (anehx) [reply addon/templates/components/work-item-button.hbs:4]: You now changed it wrong for the second time. Just think a moment about what would make sense before you change it the next time :wink:

**Comment by anehx** [comment]
Ok, now we've got basic API documentation. However, I meant user documentation. Add a page for buttons or similar where you explain the usage in proper sentences and examples like [here](https://docs.caluma.io/ember-caluma/docs/queries).

**Comment by Yelinz** [comment]
> Ok, now we've got basic API documentation. However, I meant user documentation. Add a page for buttons or similar where you explain the usage in proper sentences and examples like [here](https://docs.caluma.io/ember-caluma/docs/queries).

Added some user documentation

File: tests/dummy/app/templates/docs/buttons.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  @mutation="complete"
+  @workItemId={{this.workItemId}}
+/>
+
+<WorkItemButton @mutation="complete" @workItemId={{this.workItemId}}>
+  Lorem Ipsum
+</WorkItemButton>
+```
+
+ The WorkItemButton component expects 2 parameters:
```
Comment (anehx) [inline tests/dummy/app/templates/docs/buttons.md:27]: Count again

File: tests/dummy/app/templates/docs/buttons.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
+# Buttons
+
+`ember-caluma` provides buttons for mutating the state of work items.
+
+## Usage
+
+To use these component call them with the required parameters.
```
Comment (anehx) [inline tests/dummy/app/templates/docs/buttons.md:7]: This sentence doesn't help at all. Delete it.

File: tests/dummy/app/templates/docs/buttons.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+</WorkItemButton>
+```
+
+ The WorkItemButton component expects 2 parameters:
+
+ - `mutation` The state mutation to be used either `complete`, `skip` or `cancel`.
+ - `workItemId` Id of the work item which should be mutated.
+ - `label` (Optional) Overwrites the button text.
+
+ Additionally there are many optional parameters for the underlying UkButton component, refer to the UkButton component documentation to see what exactly they do:
```
Comment (anehx) [inline tests/dummy/app/templates/docs/buttons.md:33]: Link the docs. Other people might not know that `UkButton` comes from `ember-uikit` ...

File: tests/dummy/app/templates/docs/queries.md [hunk 15]
```diff
@@ -46,8 +46,8 @@ public methods:
 
 | Name           | Type              | Description                                                  |
 | -------------- | ----------------- | ------------------------------------------------------------ |
-| `filter`       | `String | Object` | According to the `gql` filter type defined in the query you're using. |
-| `order`        | `String | Object` | According to the `gql` order type defined in the query you're using. |
+| `filter`       | `String` | `Object` | According to the `gql` filter type defined in the query you're using. |
+| `order`        | `String` | `Object` | According to the `gql` order type defined in the query you're using. |
```
Comment (anehx) [inline tests/dummy/app/templates/docs/queries.md:50]: This was correct. Why did you change it?

File: tests/dummy/app/templates/docs/buttons.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+## Usage
+
+To use these component call them with the required parameters.
+
+### WorkItemButton
+
+The text of the button is per default the mutation name, it can be overwritten with either the `label` parameter or using the component in block form.
+
+```hbs
+<WorkItemButton @mutation="complete" @workItemId={{this.workItemId}}/>
```
Comment (anehx) [inline tests/dummy/app/templates/docs/buttons.md:14]: Don't use a variable in a documentation. No one knows what `{{this.workItemId}}` is supposed to be. Is it a UUID, is it a graphql ID? Just put it directly into the template..

**Review by anehx — CHANGES_REQUESTED** [review]

File: tests/dummy/app/templates/docs/queries.md (hunk 15 above)
Comment (Yelinz) [reply tests/dummy/app/templates/docs/queries.md:50]: Didn't seem correct to me, the words weren't formatted with the backticks

Comment (anehx) [reply tests/dummy/app/templates/docs/queries.md:50]: Thats the correct notation in JSDoc for a type that can be a `String` *or* an `Object`

Comment (Yelinz) [reply tests/dummy/app/templates/docs/queries.md:50]: image
But it gets interpreted as a line of the table and cuts off the last column of text. My solution is also not correct, but thats the issue I found with that notation.

Comment (anehx) [reply tests/dummy/app/templates/docs/queries.md:50]: Ah, just escape it then: \`String\|Object\`

File: tests/dummy/app/templates/docs/buttons.md (hunk 12 above)
Comment (Yelinz) [reply tests/dummy/app/templates/docs/buttons.md]: Changed the wording

File: tests/dummy/app/templates/docs/buttons.md (hunk 13 above)
Comment (Yelinz) [reply tests/dummy/app/templates/docs/buttons.md]: Removed

File: tests/dummy/app/templates/docs/buttons.md (hunk 14 above)
Comment (Yelinz) [reply tests/dummy/app/templates/docs/buttons.md]: Added a link

File: tests/dummy/app/templates/docs/queries.md (hunk 15 above)
Comment (Yelinz) [reply tests/dummy/app/templates/docs/queries.md]: Changed

File: addon/components/work-item-button.js [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+   * @argument {String} mutation
+   */
+  /**
+   * The id of the work item
+   * @argument {String} workItemId
+   */
+
+  @queryManager apollo;
+
+  @service intl;
```
Comment (anehx) [inline addon/components/work-item-button.js:32]: This is not used anywhere

File: addon/components/work-item-button.js [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+  cancelWorkItemMutation = cancelWorkItem;
+  completeWorkItemMutation = completeWorkItem;
+  skipWorkItemMutation = skipWorkItem;
+
+  @dropTask
+  *mutate() {
+    yield this.apollo.mutate({
+      mutation: this[`${this.args.mutation}WorkItemMutation`],
+      variables: { id: this.args.workItemId },
+    });
```
Comment (anehx) [inline addon/components/work-item-button.js:43]: Let's implement hooks so the user can provide his own feedback. Something like `@onError` and `@onSuccess` would be a good start. They need to be optional of course

File: addon/templates/components/work-item-button.hbs [hunk 19]
```diff
@@ -0,0 +1,19 @@
+<UkButton
+  ...attributes
+  @type={{@type}}
+  @disabled={{or @disabled @loading this.mutate.isRunning}}
+  @active={{@active}}
+  @loading={{or @loading this.mutate.isRunning}}
+  @color={{or @color "default"}}
```
Comment (anehx) [inline addon/templates/components/work-item-button.hbs:7]: Isn't `"default"` the default color for an `UkButton`? That would make the `or` obsolete

**Review by anehx — CHANGES_REQUESTED** [review]

File: addon/templates/components/work-item-button.hbs (hunk 19 above)
Comment (Yelinz) [reply addon/templates/components/work-item-button.hbs:7]: If we pass `@color` but it has no value it sets the color to null/ undefined overwriting the default of UkButton and having no style.

File: addon/components/work-item-button.js (hunk 17 above)
Comment (Yelinz) [reply addon/components/work-item-button.js]: Removed

File: addon/components/work-item-button.js (hunk 18 above)
Comment (Yelinz) [reply addon/components/work-item-button.js]: Added them

**Review by anehx — APPROVED** [review]
